src/core/context.ml

```diff
@@ -194,6 +194,72 @@ let rec iter_rev (ctx : 'a LF.ctx) (f : 'a LF.ctx -> 'a -> unit) : unit =
 let iter_rev' (ctx : 'a LF.ctx) (f : 'a -> unit) : unit =
   iter_rev ctx (fun _ -> f)
 
+let find_with_index (ctx : 'a LF.ctx) (f : 'a LF.ctx -> 'a -> bool) : ('a * int) option =
+  let rec go (ctx : 'a LF.ctx) (idx : int) =
+    match ctx with
+    | Empty -> None
+    | Dec (ctx', x) ->
+       begin
+         match go ctx' (idx + 1) with
```

Comment:
Consider using a higher-order function from `Support.Maybe` to eliminate this pattern matching. In particular we have an alternation operator similar to Haskell's `<|>`.